### PR TITLE
[perfdhcp] [bugfix] use the same transaction id for the duration of a DORA session.

### PR DESCRIPTION
**Pull request by pallotron**
Found this out when making some benchmark testing. The XID is incremented wrongly.
A XID should last for the duration of a SORA 4-way handshake...

```
    $ perfdhcp -4 -r 1 -n 2 -p 60 -t 1 -b mac=90:e2:ba:95:b9:28 10.88.45.33
    $ sudo tshark udp port 67
    Running as user "root" and group "root". This could be dangerous.
    Capturing on 'eth0'
      1   0.000000 10.35.166.47 -> 10.88.45.33  DHCP 295 DHCP Discover - Transaction ID 0x0
      2   0.030098  10.88.45.33 -> 10.35.166.47 DHCP 416 DHCP Offer    - Transaction ID 0x0
      3   0.030426 10.35.166.47 -> 10.88.45.33  DHCP 307 DHCP Request  - Transaction ID 0x1
      4   0.057158  10.88.45.33 -> 10.35.166.47 DHCP 416 DHCP ACK      - Transaction ID 0x1
      5   1.000689 10.35.166.47 -> 10.88.45.33  DHCP 295 DHCP Discover - Transaction ID 0x2
      6   1.027757  10.88.45.33 -> 10.35.166.47 DHCP 416 DHCP Offer    - Transaction ID 0x2
      7   1.027866 10.35.166.47 -> 10.88.45.33  DHCP 307 DHCP Request  - Transaction ID 0x3
      8   1.047429  10.88.45.33 -> 10.35.166.47 DHCP 416 DHCP ACK      - Transaction ID 0x3
```

My PR fixes this:

 ```
 $ perfdhcp -4 -r 1 -n 2 -p 60 -t 1 -b mac=90:e2:ba:95:b9:28 10.88.45.33
 $ sudo tshark udp port 67
 Running as user "root" and group "root". This could be dangerous.
 Capturing on 'eth0'
   1   0.000000 10.35.166.47 -> 10.88.45.33  DHCP 304 DHCP Discover - Transaction ID 0x0
   2   0.031211  10.88.45.33 -> 10.35.166.47 DHCP 425 DHCP Offer    - Transaction ID 0x0
   3   0.031517 10.35.166.47 -> 10.88.45.33  DHCP 316 DHCP Request  - Transaction ID 0x0
   4   0.061698  10.88.45.33 -> 10.35.166.47 DHCP 425 DHCP ACK      - Transaction ID 0x0
   5   1.000589 10.35.166.47 -> 10.88.45.33  DHCP 304 DHCP Discover - Transaction ID 0x1
   6   1.018230  10.88.45.33 -> 10.35.166.47 DHCP 425 DHCP Offer    - Transaction ID 0x1
   7   1.018360 10.35.166.47 -> 10.88.45.33  DHCP 316 DHCP Request  - Transaction ID 0x1
   8   1.058186  10.88.45.33 -> 10.35.166.47 DHCP 425 DHCP ACK      - Transaction ID 0x1
 ```

Need to modify unit tests

### CODE REVIEW

**Comment by msiodelski** [comment]
Hi Angelo,

Thanks for your pull request. There seem to be two issues with the proposed changes:
1) There are two variants of the TestControl::sendRequest4 method, and only one of them has been updated.
2) According to RFC3315 (DHCPv6), section 15.1:

" The "transaction-id" field holds a value used by clients and servers
   to synchronize server responses to client messages.  A client SHOULD
   generate a random number that cannot easily be guessed or predicted
   to use as the transaction ID for each new message it sends."

which sounds to me that the transaction id should rather be generated unless the client is retransmitting. Though, perfdhcp doesn't retransmit.

Marcin

**Comment by pallotron** [comment]
Thanks for your reply.

In theory (from what I can see in pcap files I take regularly in our prod env) the transaction ID should be generated by the client when sending DISCOVER/SOLICIT, and should be used for the duration of the 4-way handshake. Which perfdhcp is not doing (it generates a different ID when sending REQUEST) according to my tests. I know that in theory it’s against RFCs to use the TXID provided by the server. But this is a test tool, I guess you can make an exception, I don’t know perfdhcp code well enough… I think generateTransid() should be only called once per session and used for the duration of the session.

We are writing a DHCP relayer/loadbalancer internally in Go to solve some issues we have, we balance things hashing on GIADDR + TXID, and we were using perfdhcp to test the solution when I found out about the current behavior.

**Comment by msiodelski** [comment]
The code, as it is know, is against the DHCPv4 spec, but adheres to DHCPv6 spec. Since this is only a test tool I don't consider it any major issue. At the same time I am easy with updating DHCPv4 part to adhere with a spec. But, since we're making DHCPv4 part of perfdhcp to adhere with a spec it would be odd to modify the DHCPv6 to not adhere with a spec. Hence, I'd suggest we simply update the DHCPv4 part and leave DHCPv6 unchanged.

**Comment by pallotron** [comment]
Ok I will modify this PR to only change DHCPv4 part and leave DHCPv6 unchanged.

File: src/bin/perfdhcp/test_control.cc [hunk 1]
```diff
@@ -1765,8 +1766,8 @@ TestControl::sendRequest4(const TestControlSocket& socket,
     // Get the second argument if multiple the same arguments specified
     // in the command line. Second one refers to REQUEST packets.
     const uint8_t arg_idx = 1;
-    // Generate new transaction id.
-    const uint32_t transid = generateTransid();
+    // Use the same transaction id as the one used in the discovery packet.
```
Comment (pallotron) [inline src/bin/perfdhcp/test_control.cc]: not that this is in the second variant of sendRequest4

File: src/bin/perfdhcp/test_control.cc [hunk 2]
```diff
@@ -1699,7 +1699,8 @@ void
 TestControl::sendRequest4(const TestControlSocket& socket,
                           const dhcp::Pkt4Ptr& discover_pkt4,
                           const dhcp::Pkt4Ptr& offer_pkt4) {
-    const uint32_t transid = generateTransid();
+    // Use the same transaction id as the one used in the discovery packet.
+    const uint32_t transid = discover_pkt4->getTransid();
```
Comment (pallotron) [inline src/bin/perfdhcp/test_control.cc]: not that this is in the fist variant of sendRequest4

**Comment by msiodelski** [comment]
I integrated your changes into our main repository. Thanks again. Closing pull request.

**Comment by pallotron** [comment]
thank you for merging this so quickly.

**Pull request closed without merging**